src/JsonApiErrorMiddleware.php

```diff
@@ -7,18 +7,73 @@
 use Psr\Http\Message\ServerRequestInterface;
 use Psr\Http\Server\MiddlewareInterface;
 use Psr\Http\Server\RequestHandlerInterface;
+use Throwable;
+use WtfPhp\JsonApiErrors\Factories\JsonApiErrorFactory;
 
 class JsonApiErrorMiddleware implements MiddlewareInterface
 {
     private ResponseFactoryInterface $responseFactory;
+    private JsonApiErrorFactory $jsonApiErrorFactory;
+    private JsonApiErrorResponseSchema $jsonApiErrorResponseSchema;
 
-    public function __construct(ResponseFactoryInterface $responseFactory)
-    {
+    public function __construct(
+        ResponseFactoryInterface $responseFactory,
+        JsonApiErrorFactory $jsonApiErrorFactory,
+        JsonApiErrorResponseSchema $jsonApiErrorResponseSchema
+    ) {
         $this->responseFactory = $responseFactory;
+        $this->jsonApiErrorFactory = $jsonApiErrorFactory;
+        $this->jsonApiErrorResponseSchema = $jsonApiErrorResponseSchema;
     }
 
+    // INFO: Currently handles only a single Throwable.
     public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
     {
-        return $handler->handle($request);
+        try {
+            return $handler->handle($request);
+        } catch (Throwable $t) {
+            return $this->buildResponse($t);
+        }
+    }
+
+    /**
+     * @param Throwable $t
+     * @return ResponseInterface
+     */
+    private function buildResponse(Throwable $t): ResponseInterface
+    {
+        // TODO NOW: If no prop is set, do I still want to set code and message?
+        // See Test: itShouldHandleAnExceptionWithoutErrorCodeAndMessage
+        $jsonApiError = $this->jsonApiErrorFactory::createFromThrowable($t);
+
+        // TODO: Make me pretty!
+        $reasonPhrase = !empty($t->getMessage()) ? $t->getMessage() : 'Internal Server Error';
+        $status = $this->isValidHttpStatus($t->getCode()) ? $t->getCode() : 500;
+
+        /** @var JsonApiErrorResponse $response */
+        $response = $this->responseFactory->createResponse($status, $reasonPhrase);
+
+        $jsonApiErrors = $this->jsonApiErrorResponseSchema->getAsJsonApiError($jsonApiError);
+
+        // TODO NOW: Has somebody an idea how to make that body prettier?
+        $response->getBody()->write($jsonApiErrors);
+        $response->withBody($response->getBody());
+
+        return $response;
+    }
+
+    /**
+     * @param string $statusCode
+     * @return bool
+     */
+    private function isValidHttpStatus(string $statusCode): bool
+    {
+        // TODO NOW: How can we improve this?
+        $statusCodes = [
```

Comment:
i think this stuff should be put in a public constant

Reply:
We will put it in a public constant for now, but Witti is thinking about writing a http-status-code-mapper-package. 😄